### PR TITLE
Terragrunt hooks

### PR DESCRIPTION
**Pull request by eak12913**
This is my stab at implementing [#377](https://github.com/gruntwork-io/terragrunt/issues/377)

Notes:
- `run_on_error`: allows users to control whether this hook should run if a previous error occurred (from a previous hook or from terraform execution)
- `run_on_error`: if you don't specify it, `false` is the default

Before hook syntax:
```    
   before_hook "copy-file-before2" {
      commands = ["apply", "plan"]
      execute = ["echo", "Hello World"]
      run_on_error = true
    }
```

After hook syntax:
```
    after_hook "after-copy-file" {
      commands = ["apply", "plan"]
      execute = ["echo", "After Hook Running"]
      run_on_error = true
    }
```

### CODE REVIEW

File: cli/cli_app.go [hunk 1]
```diff
@@ -223,6 +224,46 @@ func runTerragrunt(terragruntOptions *options.TerragruntOptions) error {
 	return runTerragruntWithConfig(terragruntOptions, terragruntConfig, false)
 }
 
+func processBeforeAfterHookActions(beforeAfterHoook []config.BeforeAfterHook, terragruntConfig *config.TerragruntConfig, terragruntOptions *options.TerragruntOptions, previousExecError bool) bool {
```
Comment (brikis98) [inline cli/cli_app.go]: `processHooks` is probably a bit simpler of a name :)

File: config/config.go [hunk 2]
```diff
@@ -63,10 +63,24 @@ func (deps *ModuleDependencies) String() string {
 	return fmt.Sprintf("ModuleDependencies{Paths = %v}", deps.Paths)
 }
 
+// BeforeAfterHook specifies terraform commands (apply/plan) and array of os commands to execute
+type BeforeAfterHook struct {
```
Comment (brikis98) [inline config/config.go]: Could just call it `Hook`.

File: config/config.go [hunk 3]
```diff
@@ -63,10 +63,24 @@ func (deps *ModuleDependencies) String() string {
...
+}
+
 // TerraformConfig specifies where to find the Terraform configuration files
 type TerraformConfig struct {
-	ExtraArgs []TerraformExtraArguments `hcl:"extra_arguments"`
-	Source    string                    `hcl:"source"`
+	ExtraArgs  []TerraformExtraArguments `hcl:"extra_arguments"`
+	Source     string                    `hcl:"source"`
+	BeforeHook []BeforeAfterHook         `hcl:"before_hook"`
+	AfterHook  []BeforeAfterHook         `hcl:"after_hook"`
```
Comment (brikis98) [inline config/config.go]: Probably `BeforeHooks` and `AfterHooks` would be a bit clearer.

File: cli/cli_app.go [hunk 4]
```diff
@@ -261,7 +302,37 @@ func runTerragruntWithConfig(terragruntOptions *options.TerragruntOptions, terra
 			return err
 		}
 	}
-	return shell.RunTerraformCommand(terragruntOptions, terragruntOptions.TerraformCliArgs...)
+
+	terragruntOptions.Logger.Println("Running terraform with: %s", terragruntOptions)
+
+	errorOccurred := true
```
Comment (brikis98) [inline cli/cli_app.go]: Rename to `beforeHooksErr`.

File: cli/cli_app.go [hunk 5]
```diff
@@ -261,7 +302,37 @@ func runTerragruntWithConfig(terragruntOptions *options.TerragruntOptions, terra
...
+		if errorOccurred || possibleError != nil {
+			//if there was an error during before hook run
+			//or if there was an error during terraform execution
+			//make sure to pass that knowledge on to after hook execution
+			processBeforeAfterHookActions(terragruntConfig.Terraform.AfterHook, terragruntConfig, terragruntOptions, true)
+		} else {
+			processBeforeAfterHookActions(terragruntConfig.Terraform.AfterHook, terragruntConfig, terragruntOptions, false)
+		}
+	}
+	return possibleError
```
Comment (brikis98) [inline cli/cli_app.go]: It took a little while to figure out the intent of the code above. Some ideas on how to make this a bit clearer.

First, as this method can now generate multiple errors, create a type to capture them:

```go
type MultiError struct {
  Errors []error
}

func NewMultiError(errs ... error) error {
  if len(errs) == 0 {
    return nil
  }

  return MultiError{Errors: errs}
} 

func (errs MultiError) Error() string {
  errorMessages := []string{}
  for _, err := range errs {
    errorMessages = append (out, err.Error())
  }

  return fmt.Sprintf("Hit multiple errors:\n%s", strings.Join(errorMessages, "\n"))
}
```

Next, extract most of this functionality into small helper methods, and collect the errors as you go:

```go
beforeHooksErr := processHooks(terragruntConfig.Terraform.BeforeHooks, terragruntOptions)
terraformCmdErr := runTerraformCommandIfNoErrors(beforeHooksErr, terragruntOptions)
afterHooksErr := processHooks(terragruntConfig.Terraform.afterHooks, terragruntOptions, beforeHooksErr, terraformCmdErr)

return NewMultiError(beforeHooksErr, terraformCmdErr, afterHooksErr)
```

The key things to notice are:

1. `processHooks` takes an optional `...` of errors that happened previous as the final params. On the first call, this list is empty; on the second call, it has errors from before hooks and the Terraform command execution. If any of these are non-nil, it internally checks the `RunOnError` setting to determine if the hook should be executed anyway.

1. The `runTerraformCommandIfNoErrors` only executes if the errors passed into it are `nil`.

File: cli/cli_app.go [hunk 6]
```diff
@@ -223,6 +224,46 @@ func runTerragrunt(terragruntOptions *options.TerragruntOptions) error {
...
+					errorOccurred = true
+				}
+			}
+		}
+	}
+
+	return errorOccurred
+}
+
+func contains(arr []string, str string) bool {
```
Comment (brikis98) [inline cli/cli_app.go]: Stupid Go and its lack of collections. We have this method in `util.ListContainsElement`.

File: cli/cli_app.go [hunk 7]
```diff
@@ -223,6 +224,46 @@ func runTerragrunt(terragruntOptions *options.TerragruntOptions) error {
...
 }
 
+func processBeforeAfterHookActions(beforeAfterHoook []config.BeforeAfterHook, terragruntConfig *config.TerragruntConfig, terragruntOptions *options.TerragruntOptions, previousExecError bool) bool {
+
+	terraformCommand := terragruntOptions.TerraformCliArgs[0]
+
+	errorOccurred := previousExecError
+
+	for i := 0; i < len(beforeAfterHoook); i++ {
+		applicableCommands := beforeAfterHoook[i].Commands
```
Comment (brikis98) [inline cli/cli_app.go]: `beforeAfterHoook` has too many o's.

File: cli/cli_app.go [hunk 8]
```diff
@@ -223,6 +224,41 @@ func runTerragrunt(terragruntOptions *options.TerragruntOptions) error {
...
+
+	terraformCommand := terragruntOptions.TerraformCliArgs[0]
+
+	errorsOccurred := []error{}
+
+	if len(hooks) > 0 {
+		terragruntOptions.Logger.Printf("Detected %d Hooks", len(hooks))
+
+		for i := 0; i < len(hooks); i++ {
+			applicableCommands := hooks[i].Commands
```
Comment (brikis98) [inline cli/cli_app.go]: Idiomatic way to do this loop in Go is:

```go
for _, hook := range hooks {
  applicableCommands := hook.Commands
}
```

File: cli/cli_app.go [hunk 9]
```diff
@@ -223,6 +224,41 @@ func runTerragrunt(terragruntOptions *options.TerragruntOptions) error {
...
+
+			if util.ListContainsElement(applicableCommands, terraformCommand) {
+				//if there's no previous error, execute command
+				//OR if a previos error DID happen AND we want to run anyways
+				//then execute.
+				//Skip execution if there was an error AND we care about errors
+				if len(previousExecError) == 0 || hooks[i].RunOnError {
```
Comment (brikis98) [inline cli/cli_app.go]: To reduce nesting, could combine these if-statements into a method and call this method in the for-loop:

```go
func shouldRunHook(hook config.Hook, terragruntOptions *options.TerragruntOptions,  previousExecErrors ... error) bool {
  return util.ListContainsElement(hook.Commands, terragruntOptions..TerraformCliArgs[0]) && (len(previousExecErrors) == 0) || hook.RunOnError)
}
```

File: cli/cli_app.go [hunk 10]
```diff
@@ -223,6 +224,41 @@ func runTerragrunt(terragruntOptions *options.TerragruntOptions) error {
...
+					terragruntOptions.Logger.Printf("Executing hook: %s", hooks[i].Name)
+					actionToExecute := hooks[i].Execute[0]
+					actionParams := hooks[i].Execute[1:]
```
Comment (brikis98) [inline cli/cli_app.go]: What does this do if `Execute` only has 1 item? Or 0?

File: cli/cli_app.go [hunk 11]
```diff
@@ -223,6 +224,47 @@ func runTerragrunt(terragruntOptions *options.TerragruntOptions) error {
...
+	errorsOccurred := []error{}
+
+	terragruntOptions.Logger.Printf("Detected %d Hooks", len(hooks))
+
+	for _, curHook := range hooks {
+		allPreviousErrors := append(previousExecError, errorsOccurred...)
+		if shouldRunHook(curHook, terragruntOptions, allPreviousErrors...) {
+			terragruntOptions.Logger.Printf("Executing hook: %s", curHook.Name)
+			actionToExecute := curHook.Execute[0]
+			if actionToExecute != "" {
```
Comment (brikis98) [inline cli/cli_app.go]: Prob better to validate this stuff during parsing, not part way through execution.

File: cli/cli_app.go [hunk 12]
```diff
@@ -223,6 +224,47 @@ func runTerragrunt(terragruntOptions *options.TerragruntOptions) error {
...
+				actionParams := curHook.Execute[1:]
```
Comment (brikis98) [inline cli/cli_app.go]: What if it's just a command with no args?

File: test/fixture-hooks/after-only/terraform.tfvars [hunk 13]
```diff
@@ -0,0 +1,12 @@
+terragrunt = {
+  terraform {
+    // source = "test"
+
+    # This hook configures Terragrunt to do a simple echo statement after executing any Terraform command
```
Comment (brikis98) [inline test/fixture-hooks/after-only/terraform.tfvars]: Comment isn't accurate

File: test/fixture-hooks/before-and-after/terraform.tfvars [hunk 14]
```diff
@@ -0,0 +1,20 @@
+terragrunt = {
+  terraform {
+    // source = "test"
+
+    # This hook configures Terragrunt to copy /foo to /bar before executing apply or plan
```
Comment (brikis98) [inline test/fixture-hooks/before-and-after/terraform.tfvars]: None of these comments seem to correspond to what the code is doing?

File: test/fixture-hooks/skip-on-error/terraform.tfvars [hunk 15]
```diff
@@ -0,0 +1,37 @@
+terragrunt = {
+  terraform {
+    // source = "test"
+
+    # This hook purposely causes an error
+    before_hook "before_hook_1" {
+      commands = ["apply", "plan"]
+      execute = ["cd","garbageFolder"]
```
Comment (brikis98) [inline test/fixture-hooks/skip-on-error/terraform.tfvars]: Could also just `exit 1`

File: test/integration_test.go [hunk 16]
```diff
@@ -60,6 +66,97 @@ func init() {
 	rand.Seed(time.Now().UnixNano())
 }
 
+func TestTerragruntBeforeHook(t *testing.T) {
+	cleanupTerraformFolder(t, TEST_FIXTURE_HOOKS_BEFORE_ONLY_PATH)
```
Comment (brikis98) [inline test/integration_test.go]: Ideally, we can call `t.Parallel()` at the top of each of these tests so they all run in parallel. Note that this only works if they all run in different folders.

Comment (brikis98) [reply test/integration_test.go]: One other catch: the `file.out` that may of the hooks create may be left overs from previous test runs. So probably, the tests should copy the folder you're testing (e.g. `TEST_FIXTURE_HOOKS_BEFORE_ONLY_PATH`) to a new temp folder each test run to ensure we're not seeing an older `file.out`. There are several helper methods in in `integration_test.go` to do that copying.

File: test/integration_test.go [hunk 17]
```diff
@@ -60,6 +66,97 @@ func init() {
...
+
+	err := runTerragruntCommand(t, fmt.Sprintf("terragrunt apply --terragrunt-non-interactive --terragrunt-working-dir %s", TEST_FIXTURE_HOOKS_SKIP_ON_ERROR_PATH), &stdout, &stderr)
+
+	output := stderr.String()
+	if err != nil {
+		assert.Contains(t, output, "BEFORE_SHOULD_DISPLAY")
+		assert.NotContains(t, output, "BEFORE_NODISPLAY")
+
+		assert.Contains(t, output, "AFTER_SHOULD_DISPLAY")
+		assert.NotContains(t, output, "AFTER_NODISPLAY")
```
Comment (brikis98) [inline test/integration_test.go]: Nice tests!

File: cli/cli_app.go (hunk 12 above)
Comment (eak12913) [reply cli/cli_app.go]: This should work. I added a test to make sure it does: `TestTerragruntBeforeOneArgAction`

File: test/integration_test.go (hunk 16 above)
Comment (eak12913) [reply test/integration_test.go]: Good point - will look into it and implement

File: test/integration_test.go (hunk 17 above)
Comment (eak12913) [reply test/integration_test.go]: Thanks :-)

File: config/config.go [hunk 18]
```diff
@@ -359,6 +374,15 @@ func convertToTerragruntConfig(terragruntConfigFromFile *terragruntConfigFile, t
 		terragruntConfig.RemoteState = terragruntConfigFromFile.RemoteState
 	}
 
+	//Validate before and after hooks to make sure that they are properly configured
+	allHooks := append(terragruntConfigFromFile.Terraform.BeforeHooks, terragruntConfigFromFile.Terraform.AfterHooks...)
+
+	for _, curHook := range allHooks {
+		if curHook.Execute[0] == "" {
```
Comment (brikis98) [inline config/config.go]: Could `curHook.Execute` be an empty list? So that this throws an out of bounds error?

File: errors/invalidargerror.go [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+	s string
+}
+
+func (e InvalidArgError) Error() string {
+	return e.s
+}
+
+func NewInvalidArgError(message string) error {
+	return InvalidArgError{s: message}
+}
```
Comment (brikis98) [inline errors/invalidargerror.go]: A couple thoughts on errors:

1. Unlike Java, you don't have to define them in their own files. We often put the errors relevant to a given bit of code in the same file (often at the bottom) as that code.
1. If your error is just a wrapper for a single type, you can declare it as an "alias" for that type to simplify usage a bit:

```go
type InvalidArgError string

func (e InvalidArgError) Error() string {
	return string(e)
}

// To create this error elsewhere in the code:
return InvalidArgError("My error message")
```

File: test/fixture-hooks/bad-arg-action/terraform.tfvars [hunk 20]
```diff
@@ -0,0 +1,11 @@
+terragrunt = {
+  terraform {
+
+    # This hook is purposely misconfigured to trigger an error
+    before_hook "before_hook_1" {
+      commands = ["apply", "plan"]
+      execute = [""]
```
Comment (brikis98) [inline test/fixture-hooks/bad-arg-action/terraform.tfvars]: Test this with `[]` (empty list is different than list with an empty string)

File: test/fixture-hooks/interpolations/terraform.tfvars [hunk 21]
```diff
@@ -0,0 +1,11 @@
+terragrunt = {
+  terraform {
+
+    # This hook echos out user's HOME path or HelloWorld 
+    before_hook "before_hook_1" {
+      commands = ["apply", "plan"]
+      execute = ["echo", "${get_env("HOME", "HelloWorld")}"]
```
Comment (brikis98) [inline test/fixture-hooks/interpolations/terraform.tfvars]: Nice :)

File: test/fixture-hooks/skip-on-error/terraform.tfvars [hunk 22]
```diff
@@ -0,0 +1,37 @@
+terragrunt = {
+  terraform {
+    // source = "test"
```
Comment (brikis98) [inline test/fixture-hooks/skip-on-error/terraform.tfvars]: Remove?

File: config/config.go (hunk 18 above)
Comment (eak12913) [reply config/config.go]: Good point - Added a new test for this and fixed the underlying issue

File: errors/invalidargerror.go (hunk 19 above)
Comment (eak12913) [reply errors/invalidargerror.go]: Ah ok

**Comment by eak12913** [comment]
Yikes! You’re right! I’ll try to get that done today or tomorrow.

**Comment by brikis98** [comment]
I'll wait for the tests to run, and if they pass, issue a new release.

**Comment by brikis98** [comment]
New release is here!

https://github.com/gruntwork-io/terragrunt/releases/tag/v0.14.4

Thanks @eak12913!

**Comment by ebarault** [comment]
Excellent, a very big thank you for bringing this to terragrunt @eak12913

**Comment by eak12913** [comment]
No problem :-) Happy to help!

**Comment by ebarault** [comment]
@eak12913 : i tried playing a bit with new hooks and i realized hooks are running inside terragrunt's tmp dir, which means after the terragrunt has created it and moved the files there.

in my case, i'd also need hooks that would happen before in the terragrunt source code dir.

what do you think would be the simpler to implement, add an option to the `before_hook` block? or create another type of hook? such as `initial_hook`?

**Comment by eak12913** [comment]
@ebarault: I discussed this with @brikis98 and he proposed an interesting idea:  What if we provided a `working_dir` parameter for the hook? This would default to terragrunt's working dir, but could be overridden to anything you need.  Could this work for you?

**Comment by ebarault** [comment]
By "terragrunt's working dir", do you mean the location where the original `terraform.tfvars` sits or the tmp dir where it is copied to?

See https://github.com/gruntwork-io/terragrunt/issues/453, i have also a temporality issue with when the hooks happen: thanks to terragrunt's interpolation functions i managed running the hooks in the source code dir, but then the problem is that they happen after terragrunt's parsing of `required_var_files` / `optional_var_files`.

So if the file (which i want to generate with hooks) is missing at terragrunt's execution, even though it's generated just after, it will be skipped by terragrunt.
```
[terragrunt] 2018/03/28 19:53:13 Skipping var-file /path/to/component.tfvars as it does not exist
```

I ended up generating the file directly in the tmp dir as suggested by @lorengordon using the `.auto.tfvars` extension so it gets automatically parsed by terraform.

Not an ideal solution if you ask me as it makes the workflow difficult to understand, and we have to care about the tmp folder which should remain out of end user's concern.

A more natural solution would have the before hooks doing any required action actually before any other terragrunt step, then the regular terragrunt workflow takes place.

**Pull request merged**